swift/ql/lib/codeql/swift/frameworks/StandardLibrary/NSData.qll

```diff
@@ -0,0 +1,92 @@
+/** Provides classes and models to work with `NSData`-related objects. */
+
+import swift
+private import codeql.swift.dataflow.DataFlow
+private import codeql.swift.dataflow.ExternalFlow
+private import codeql.swift.dataflow.FlowSteps
+
+/** The class `NSData`. */
+class NsData extends ClassDecl {
+  NsData() { this.getFullName() = "NSData" }
+}
+
+/** The class `NSMutableData`. */
+class NsMutableData extends ClassDecl {
+  NsMutableData() { this.getFullName() = "NSMutableData" }
+}
+
+private class NsDataSources extends SourceModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        ";NSData;true;init(contentsOf:);;;ReturnValue;remote",
+        ";NSData;true;init(contentsOf:options:);;;ReturnValue;remote"
+      ]
+  }
+}
+
+private class NsDataSummaries extends SummaryModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        ";NSData;true;init(bytes:length:);;;Argument[0];ReturnValue;taint",
+        ";NSData;true;init(bytesNoCopy:length:);;;Argument[0];ReturnValue;taint",
+        ";NSData;true;init(bytesNoCopy:length:deallocator:);;;Argument[0];ReturnValue;taint",
+        ";NSData;true;init(bytesNoCopy:length:freeWhenDone:);;;Argument[0];ReturnValue;taint",
+        ";NSData;true;init(data:);;;Argument[0];ReturnValue;taint",
+        ";NSData;true;init(contentsOfFile:);;;Argument[0];ReturnValue;taint",
+        ";NSData;true;init(contentsOfFile:options:);;;Argument[0];ReturnValue;taint",
+        ";NSData;true;init(contentsOf:);;;Argument[0];ReturnValue;taint",
+        ";NSData;true;init(contentsOf:options:);;;Argument[0];ReturnValue;taint",
+        ";NSData;true;init(contentsOfMappedFile:);;;Argument[0];ReturnValue;taint",
+        ";NSData;true;init(base64Encoded:options:);;;Argument[0];ReturnValue;taint",
+        ";NSData;true;init(base64Encoding:);;;Argument[0];ReturnValue;taint",
+        ";NSData;true;base64EncodedData(options:);;;Argument[-1];ReturnValue;taint",
+        ";NSData;true;base64EncodedString(options:);;;Argument[-1];ReturnValue;taint",
+        ";NSData;true;base64Encoding();;;Argument[-1];ReturnValue;taint",
+        ";NSData;true;dataWithContentsOfMappedFile(_:);;;Argument[0];ReturnValue;taint",
+        // TODO: Needs block flow
+        // ";NSData;true;enumerateBytes(_:);;;Argument[-1];Argument[0].Parameter[0];taint"
```

Comment:
I think flow through [closures](https://docs.swift.org/swift-book/LanguageGuide/Closures.html) (which I called "blocks" here) is a general data flow feature that should be handled separately (should be equivalent to lambda flow in other languages). Probably related to "flow through `withUnsafeBytes` and similar methods", which we already have an issue for. /cc @MathiasVP in case you have some plans for this already.

Reply:
I _thought_ we had support for flow-through blocks (see [here](https://github.com/github/codeql/blob/1eec067474c45397a2453925422786c21ca27dac/swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll#L644)). But we might be missing something to connect this to MaD 🤔? The taint source I added [here](https://github.com/github/codeql/pull/10699/files#diff-6a9b9f57af2b6dc32149efa7893028ee8aa080b7f5eae35d3c797d30af604d14R67) convinced me it was working for MaD, though.

Reply:
Huh, I don't know how I previously tested that but it indeed works after uncommenting the CSV row (see 2ac06b8). That's great, we have better MaD support than I thought! Sorry for unnecessarily bothering you @MathiasVP, and thanks for pointing this out @geoffw0 🙏

Reply:
Great, I'm just happy it's working!